proxy/grpcproxy/cluster.go

```diff
@@ -15,38 +15,177 @@
 package grpcproxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 
 	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	gnaming "google.golang.org/grpc/naming"
 )
 
 type clusterProxy struct {
-	client *clientv3.Client
+	clus clientv3.Cluster
+	ctx  context.Context
+	gr   *naming.GRPCResolver
+
+	// self is true to return its own client listen URLs
+	// for member list API
+	self  bool
+	laddr string
+
+	wa   gnaming.Watcher
+	umu  sync.RWMutex
+	umap map[string]struct{}
+}
+
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// 'self' is true then member list API returns the client listen addresses.
+// The returned channel is closed when there is grpc-proxy endpoint registered
+// and the client's context is canceled so the 'register' loop returns.
+func NewClusterProxy(c *clientv3.Client, self bool, laddr string, prefix string) (pb.ClusterServer, <-chan struct{}, error) {
+	if self && prefix != "" {
+		return nil, nil, fmt.Errorf("both 'self' and 'prefix' are configured")
+	}
+	donec := make(chan struct{})
+
+	cp := &clusterProxy{
+		clus: c.Cluster,
+		ctx:  c.Ctx(),
+		gr:   &naming.GRPCResolver{Client: c},
+
+		self:  self,
+		laddr: laddr,
+	}
+	if prefix == "" {
+		close(donec)
+		return cp, donec, nil
+	}
+
+	go cp.resolve(prefix, donec)
+	return cp, donec, nil
+}
+
+// allow maximum 1 retry per second
+const resolveRetryRate = 1
+
+func (cp *clusterProxy) resolve(prefix string, donec chan struct{}) {
+	rm := rate.NewLimiter(rate.Limit(resolveRetryRate), resolveRetryRate)
+	for rm.Wait(cp.ctx) == nil {
+		watcher, err := cp.gr.Resolve(prefix)
+		if err != nil {
+			plog.Warningf("failed to resolve %q (%v)", prefix, err)
+			continue
+		}
+		cp.wa = watcher
+		break
+	}
+
+	// pass 'donec' to signal monitor routine return
+	go cp.monitor(donec)
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
-		client: c,
+func (cp *clusterProxy) monitor(donec chan struct{}) {
+	// to signal monitor routine return
+	defer close(donec)
+
+	for cp.ctx.Err() == nil {
+		ups, err := cp.wa.Next()
+		if err != nil {
+			plog.Warningf("clusterProxy watcher error (%v)", err)
+			if grpc.ErrorDesc(err) == naming.ErrWatcherClosed.Error() {
+				return
+			}
+		}
+
+		cp.umu.Lock()
+		if len(cp.umap) == 0 {
+			cp.umap = make(map[string]struct{}, len(ups))
```

Comment:
Initialize `umap` up front in `NewClusterProxy` if the resolver / self resolution is configured instead of having this special case in the update logic?